### PR TITLE
Enables entity setting and config reloading

### PR DESCRIPTION
**Pull request by tastybento**
Tweaks to config.yml to be more cave-y.

Entities are set to not despawn. Probabilities of materials and entities
can have decimals and be sub 1% if required. For some entities, 1% was
too high a rate.

Adds a debug setting in config.yml that will enable you to see where
entities have been spawned. It can be removed at some point.

### CODE REVIEW

File: src/main/java/world/bentobox/caveblock/Settings.java [hunk 1]
```diff
@@ -26,1898 +32,1894 @@
...
+
+
+    /**
+     * This method returns the seaHeight object.
+     * @return the seaHeight object.
+     */
+    @Override
+    public int getSeaHeight()
+    {
+        return 0;
```
Comment (tastybento) [inline src/main/java/world/bentobox/caveblock/Settings.java]: The chunk generator doesn't support this so I just removed it.

File: src/main/java/world/bentobox/caveblock/Settings.java [hunk 2]
```diff
@@ -26,1898 +32,1894 @@
...
+
+
+    /**
+     * This method returns the numberOfBlockGenerationTries value.
+     * @return the value of numberOfBlockGenerationTries.
+     */
+    public int getNumberOfBlockGenerationTries()
+    {
+        return numberOfBlockGenerationTries;
+
```
Comment (tastybento) [inline src/main/java/world/bentobox/caveblock/Settings.java]: This is just so you can find out where the entities are spawning so you can check them out by teleporting there.

File: src/main/java/world/bentobox/caveblock/generators/populators/EntitiesPopulator.java [hunk 3]
```diff
@@ -65,20 +64,16 @@ public EntitiesPopulator(CaveBlock addon)
...
-                    {
-                        int x = random.nextInt(15);
-                        int z = random.nextInt(15);
-                        int y = Math.min(worldHeight - 2, subY + random.nextInt(15));
-
-                        this.tryToPlaceEntity(world, chunk.getBlock(x, y, z), entry.getKey(), x, z, chances.get(world.getEnvironment()).mainMaterial);
-                    }
+                    int y = Math.min(worldHeight - 2, subY + random.nextInt(15));
+                    // Spawn only in middle of chunk because bounding box will grow out from here
+                    this.tryToPlaceEntity(world, chunk.getBlock(7, y, 7), entry.getKey(), chances.get(world.getEnvironment()).mainMaterial);
```
Comment (tastybento) [inline src/main/java/world/bentobox/caveblock/generators/populators/EntitiesPopulator.java]: This is a simplification so it means the positioning is less random, but I don't think it'll matter too much.

File: src/main/java/world/bentobox/caveblock/generators/populators/EntitiesPopulator.java [hunk 4]
```diff
@@ -111,148 +106,53 @@ public void populate(World world, Random random, Chunk chunk)
             if (entity != null)
             {
                 entityMap.put(entity,
-                        new Pair<>(Integer.parseInt(splitString[2]), Integer.parseInt(splitString[3])));
+                        new Pair<>(Double.parseDouble(splitString[2]), Integer.parseInt(splitString[3])));
```
Comment (tastybento) [inline src/main/java/world/bentobox/caveblock/generators/populators/EntitiesPopulator.java]: Enables probability to be sub 1%

File: src/main/java/world/bentobox/caveblock/generators/populators/EntitiesPopulator.java [hunk 5]
```diff
@@ -111,148 +106,53 @@ public void populate(World world, Random random, Chunk chunk)
...
+            // Make space for entity based on the entity's size
+            BoundingBox bb = e.getBoundingBox();
+            for (int x = (int) bb.getMinX(); x < bb.getMaxX(); x++) {
+                for (int z = (int) bb.getMinZ(); z < bb.getMaxZ(); z++) {
+                    int y = (int) bb.getMinY();
+                    Block b = world.getBlockAt(x, y, z);
+                    for (; y < bb.getMaxY(); y++) {
+                        if (addon.getSettings().isDebug()) {
+                            addon.log("DEBUG: Entity spawn: " + world.getName() + " " + x + " " + y + " " + z + " " + e.getType());
+                        }
```
Comment (tastybento) [inline src/main/java/world/bentobox/caveblock/generators/populators/EntitiesPopulator.java]: We can pull this code out once we are happy with the spawning.

File: src/main/java/world/bentobox/caveblock/generators/populators/MaterialPopulator.java [hunk 6]
```diff
@@ -62,99 +62,82 @@ public MaterialPopulator(CaveBlock addon)
...
-     * @return true, if all chunks around given block are generated.
-     */
-    private boolean isValidBlock(World world, Block block, int x, int z)
-    {
-        return x > 0 && x < 15 && z > 0 && z < 15 ||
-                world.isChunkGenerated(block.getX() + 1, block.getZ()) &&
-                world.isChunkGenerated(block.getX() - 1, block.getZ()) &&
-                world.isChunkGenerated(block.getX(), block.getZ() - 1) &&
-                world.isChunkGenerated(block.getX(), block.getZ() + 1);
-    }
```
Comment (tastybento) [inline src/main/java/world/bentobox/caveblock/generators/populators/MaterialPopulator.java]: If the original coordinates are not on the edge of the chunk then this code should not be needed. Apparently vanilla never mods any blocks on chunk boundaries for this reason.

Comment (BONNe) [reply src/main/java/world/bentobox/caveblock/generators/populators/MaterialPopulator.java]: world.isChunkGenerated() allowed to do it, if chunk near is generated.

**Pull request merged**